Fix bug that prevented --malloc-may-fail from working

hannes-steffenhagen-diffblue opened the pull request:
We have code in our stdlib models that depends on two global variables,
__CPROVER_malloc_may_fail and __CPROVER_malloc_failure_mode. Previously
these were set based on the ansi-c config in builtin_additions by
generating C code containing definitions for these variables.

The problem with that is that if you compile an executable using goto-cc
we also load the builtin_additions, so this executable already contains
definitions for these variables which can't be overridden.

This works around this issue by instead declaring these variables as
extern in builtin_additions, and setting their values dynamically as
last-minute instrumentation at the end of __CPROVER_initialize.  Giving
these no/nondet initial values is important because otherwise
preprocessing like constant folding on goto-binaries before cbmc
actually runs could remove branches we can, in fact, take with different
values set by --malloc-fail-* flags.

An alternative would be to instead dynamically change the code for the
models functions as instrumentation, but this has a couple of problems;
For instance this means we might change the behaviour of custom stdlib
models (which we allow via --no-library) or custom goto instrumentation
in unexpected ways. I believe this way is the least likely to cause any
friction like that.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

hannes-steffenhagen-diffblue committed:
Fix bug that prevented --malloc-may-fail from working

We have code in our stdlib models that depends on two global variables,
__CPROVER_malloc_may_fail and __CPROVER_malloc_failure_mode. Previously
these were set based on the ansi-c config in builtin_additions by
generating C code containing definitions for these variables.

The problem with that is that if you compile an executable using goto-cc
we also load the builtin_additions, so this executable already contains
definitions for these variables which can't be overridden.

This works around this issue by instead declaring these variables as
extern in builtin_additions, and setting their values dynamically as
last-minute instrumentation at the end of __CPROVER_initialize.  Giving
these no/nondet initial values is important because otherwise
preprocessing like constant folding on goto-binaries before cbmc
actually runs could remove branches we can, in fact, take with different
values set by --malloc-fail-* flags.

An alternative would be to instead dynamically change the code for the
models functions as instrumentation, but this has a couple of problems;
For instance this means we might change the behaviour of custom stdlib
models (which we allow via --no-library) or custom goto instrumentation
in unexpected ways. I believe this way is the least likely to cause any
friction like that.

peterschrammel commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -186,17 +186,15 @@ void ansi_c_internal_additions(std::string &code)
 
-    "int " CPROVER_PREFIX "malloc_failure_mode="+
-      std::to_string(config.ansi_c.malloc_failure_mode)+";\n"
+    "extern int " CPROVER_PREFIX "malloc_failure_mode;\n"
```
Can the malloc-X options be removed from goto-cc now?

hannes-steffenhagen-diffblue replied:
@peterschrammel we never added them to `goto-cc`. What we did add to `goto-cc` was `--object-bits`, which could be removed with a similar approach (or by doing the check the object bits are used for in a different way entirely the way @danpoe was suggesting).

peterschrammel commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -778,6 +779,47 @@ bool cbmc_parse_optionst::set_properties()
+/// Some variables have different initial values based on what flags are being passed to cbmc;
+/// since the _start function might've been generated by goto-cc before getting passed over here
+/// we need to add the initialisation of these variables here as a post-processing step
+static void
```
I'm not sure this code should be in this class - better move it into a separate file in goto-programs or goto-instrument.

hannes-steffenhagen-diffblue replied:
Done

martin-cs commented on src/cbmc/cbmc_parse_options.cpp:
```diff
@@ -879,6 +880,8 @@ bool cbmc_parse_optionst::process_goto_program(
   link_to_library(
     goto_model, log.get_message_handler(), cprover_c_library_factory);
 
+  add_malloc_may_fail_variable_initializations(goto_model);
```
If this is now part of the correct set up for goto programs can you either A. add this to all such set-up code ( goto-analyzer, goto-diff, goto-instrument at least) or B. add it to one of the existing methods -- I would not oppose it being part of `link_to_library` as it is part of the handling of the library.

hannes-steffenhagen-diffblue replied:
It definitely can’t go into `link_to_library` and shouldn’t go into any of the instrumentation tools (i.e. it can’t go into `link_to_library` _because_ it shouldn’t go into the instrumentation tools). The reason this is put in this location is specifically to _avoid_ instrumentation code setting these variables, they should _only_ be set by cbmc. Arguably jbmc and goto-analyzer too, but _definitely_ not goto-instrument etc.

The thing with `goto-analyzer` though is also that because the stdlib models are cbmc models, I don't think they'd even make sense for goto-analyzer (and I don't think they're loaded). Similar story with jbmc, I don't think it loads the C models (that would make more sense like to verify C code interacting with java code I guess, but I don't know if anyone's actually doing that).

hannes-steffenhagen-diffblue replied:
To summarise above, I don’t really agree from what I understand of the situation, but maybe I’m missing something.

martin-cs replied:
They shouldn't just be `cbmc` models; `goto-analyzer` will need them as well.  They may not work that well at the moment but let's not make them any more broken than they currently are.  In `jbmc` I don't know what their models are written in, Java or C.  It would be good if it was possible for it to work.

My concern is avoiding divergence in how the tools set up and configure goto programs.  This has been a source of subtle bugs in the past.

hannes-steffenhagen-diffblue replied:
@martin-cs Again, I _could_ put this code in goto-analyzer but I still don’t quite understand what good it’ll do. The models library as a matter of fact is currently not being used for goto-analyzer (and as far as I know this has been the case for _at least_ 5 years), so if I add this code it’ll naturally go untested. So the problem is that then I’m adding code to goto-analyzer that is vestigial to begin with (i.e. literally a no-op in all circumstances). While I do understand that we want to keep goto transforms consistent between different tools in _this_ particular case I don’t see the value in it. Right now goto-analyzer  doesn’t even understand things like `__CPROVER_allocate` etc (or rather, I don’t think any of the domains do). For me the problem here is I’m adding code to places that I _know_ does nothing right now, and that I can’t be sure will do the right thing whenever someone changes the surrounding context in such a way that would allow it to do something. Whenever someone in the future would read that goto-analyzer code and asks "what is this bit of the code for", the true answer would be "it does absolutely nothing because the preconditions for it doing anything are not there". If it ever starts doing anything we’ll have no way of knowing it’s doing the right thing because we didn’t (and couldn’t) test it until then.

I think we’re both concerned about maintainability here, but we seem to disagree on which way round is better for maintainability.

@peterschrammel @piotr-grabalski @NlightNFotis @thomasspriggs would you mind chipping in? Very much willing to be overruled here, but I want to be sure we’re doing the right thing here.

hannes-steffenhagen-diffblue replied:
@martin-cs Would you be OK with it if I made a different PR adding these changes to goto-analyzer and we can have the discussion on whether or not it _should_ go there on that PR? Getting this fix in for cbmc at least is somewhat time critical for us.

peterschrammel replied:
Yes, I think it should be called by goto-analyzer too.
IMO, this shouldn't be implemented through the models library at all, but by a pass over the goto model - but that may follow afterwards. This PR is for urgently fixing a major bug in CBMC when used with goto-cc. I.e. priority is getting the bug fixed today. Tomorrow we can look how to improve the solution.

hannes-steffenhagen-diffblue replied:
The main reason I’d want it to be in the models library is because IMHO whether or not this can return NULL is part of the model, and you should be able to figure out what behaviour the models library has from looking at the library code. Instrumenting logic means distributing the logic between the models library source files and different parts of the C++ code base, but like you said we should discuss how to move forward with this when we’ve actually fixed the problem.

martin-cs replied:
@hannes-steffenhagen-diffblue would you be OK adding a comments (and commented out code?) in the other locations saying why it shouldn't / isn't / may be needed?  Would be happy to approve that and get it merged in an hour or so when the CI passes?

@peterschrammel I feel the library solution has a number of advantages; should we talk about this at some point?

thomasspriggs replied:
:thinking: There is kind of a design decision to be made, as to whether these checks should be part of the model or to be an instrumentation pass. I liked the idea of this being an instrumentation pass.

:pick: If we are sticking with them being part of the implementation of the model, then I think the approach outlined in this PR isn't good in the long term either. The reason being that this is separating correctly configuring the model/library from adding it. This leaves things in an invalid state in between the two stages, where it is added but not correctly configured. However I am willing to approve as is because this is the simplest way to keep the existing user interface and fix the issue in a short period of time.

martin-cs replied:
@thomasspriggs agreed -- we do need to discuss this.  Adding the models and configuring them are two different things and yes, we do need to think about it.  Do you want to open an issue to discuss this?

hannes-steffenhagen-diffblue committed:
Fix constant_propagation tests

These constant_propagation tests expect specific numbers of assignments.
Since we changed around how we initialize the variables for the
--malloc-may-fail feature these numbers have changed.

thomasspriggs commented on src/goto-programs/add_malloc_may_fail_variable_initializations.cpp:
```diff
@@ -0,0 +1,62 @@
+
+  initialize_function.body.insert_before(
+    initialize_function_end,
+    goto_programt::make_assignment(make_integer_assignment(
```
:pick: I think I would have replaced the existing assignment in cprover_initialise, rather than adding extra assignments. Your approach will make the resulting code in `cprover_initialise` more difficult to read, when it is printed for debugging purposes. I think that replacing the existing assignment would also have reduced the need to modify the regression tests.

hannes-steffenhagen-diffblue replied:
@thomasspriggs I don’t understand how having an extra assignment is more difficult to read/debug than adding extra logic to functions at runtime

hannes-steffenhagen-diffblue committed:
Add commented out code relating to --malloc-may-fail

Also add some explanation for why it’s commented out

martin-cs reviewed (approved):
Thanks.

If possible, a comment in `goto-instrument` saying "Don't do this" and something for `goto-diff` ( @peterschrammel : what is `goto-diff`'s status) would be great but those could be in another PR if you need to get this moving ASAP.

thomasspriggs commented on jbmc/src/jbmc/jbmc_parse_options.cpp:
```diff
@@ -726,6 +726,10 @@ int jbmc_parse_optionst::get_goto_program(
 
+    // if we added the ansi-c library models here this should also call
+    // add_malloc_may_fail_variable_initializations(goto_model);
+    // here
```
:no_entry_sign: I am not sure this comment is helpful. JBMC uses its own models rather than the ansi-c models. Also the java language uses `new`, not malloc. This is parsed from the class file into `ID_java_new` / `ID_java_new_array`, which is then lowered directly into `ID_allocate`. This bypasses the use of a malloc function entirely. See `jbmc/src/java_bytecode/remove_java_new.cpp` for details.

hannes-steffenhagen-diffblue replied:
@thomasspriggs there are valid reasons why you might want to add the ansi-c library here; e.g. if there was JNI support and you wanted to link your java models to C models. Not a thing as of now as far as I know, but it's not inconceivable that his would be a concern in the future.

thomasspriggs replied:
Ok, that make a little more sense, if you were thinking about JNI support. AFAIK JNI support isn't planned for the near future. I also suspect that it would be tricky to implement given the complexities of lazy loading.

hannes-steffenhagen-diffblue replied:
Yeah, that’s why I didn’t want to put it here in the first place! But if you wanted to load the C library, you’d have to add this code and the flags as well, which I believe was @martin-cs's concern and why he asked for these comments.

martin-cs replied:
@thomasspriggs : Yes, as @hannes-steffenhagen-diffblue observed JNI is one thing.  Another is that there are a lot of implicit assumptions in the modelling and instrumentation that the C library is present.  It is likely that future modelling of more sophisticated things may well make use of / depend on the C library and I would hate to see the intermediate representation diverge because of this.

@hannes-steffenhagen-diffblue : yes, my concern is in divergence of how we handle things and the "well it works if you use X but not if you use Y first, then X".  This way the divergence is clear and documented in the relevant place (ideally it would be in `goto-instrument` and `goto-diff` too).

chrisr-diffblue reviewed (approved):
Adding my approval to the others. Though I agree it looks like some broader/longer term discussion is also needed.
